pxtblocks/blocklyloader.ts

```diff
@@ -2269,6 +2273,268 @@ namespace pxt.blocks {
+                option.callback = function() {
+                let def = Blockly.Procedures.getDefinition(name, workspace);
+                    def && def.select();
+                };
+                options.push(option);
+            },
+            defType_: 'procedures_defnoreturn'
+        }
+
+        Blockly.Procedures.flyoutCategory = function (workspace: Blockly.Workspace) {
+            let xmlList: HTMLElement[] = [];
+
+            const newFunction = lf("Make a Function");
+            const newFunctionTitle = lf("New function name:");
+
+            // Add the "Make a function" button
+            let button = goog.dom.createDom('button');
+            button.setAttribute('text', newFunction);
+            button.setAttribute('callbackKey', 'CREATE_FUNCTION');
+
+            let createFunction = (name: string) => {
+                /**
+                 * Create matching definition block.
+                 * <xml>
+                 *   <block type="procedures_defreturn" x="10" y="20">
+                 *     <field name="NAME">test</field>
+                 *   </block>
+                 * </xml>
+                 */
+                let topBlock = workspace.getTopBlocks(true)[0];
+                let x = 0, y = 0;
+                if (topBlock) {
+                    let xy = topBlock.getRelativeToSurfaceXY();
+                    x = xy.x + (Blockly as any).SNAP_RADIUS * (topBlock.RTL ? -1 : 1);
+                    y = xy.y + (Blockly as any).SNAP_RADIUS * 2;
+                }
+                let xml = goog.dom.createDom('xml');
+                let block = goog.dom.createDom('block');
+                block.setAttribute('type', 'procedures_defnoreturn');
+                block.setAttribute('x', String(x));
+                block.setAttribute('y', String(y));
+                let field = goog.dom.createDom('field');
+                field.setAttribute('name', 'NAME');
+                field.appendChild(document.createTextNode(name));
+                block.appendChild(field);
+                xml.appendChild(block);
+                let newBlockIds = Blockly.Xml.domToWorkspace(xml, workspace);
+                // Close flyout and highlight block
+                (workspace as any).toolbox_.clearSelection();
+                let newBlock = workspace.getBlockById(newBlockIds[0]);
+                newBlock.select();
+            }
+
+            workspace.registerButtonCallback('CREATE_FUNCTION', function(button) {
+                let promptAndCheckWithAlert = (defaultName: string) => {
+                    Blockly.prompt(newFunctionTitle, defaultName, function(newFunc) {
+                        // Merge runs of whitespace.  Strip leading and trailing whitespace.
+                        // Beyond this, all names are legal.
+                        if (newFunc) {
+                            newFunc = newFunc.replace(/[\s\xa0]+/g, ' ').replace(/^ | $/g, '');
+                            if (newFunc == newFunction) {
```

Comment:
should you check against the dictionary of reserved keywords? (can you name a function "var" or "function" ?)